### PR TITLE
Support scorecards

### PR DESCRIPTION
**Pull request by taimoor-at-opslevel**
https://github.com/OpsLevel/team-platform/issues/33

Relies on https://github.com/OpsLevel/opslevel-go/pull/233

## Tophatting

Step 1: wrote terraform file

```
resource "opslevel_team" "platform_team" {
    name = "platform"
}

resource "opslevel_filter" "rds_filter" {
  name = "rds apps"
  predicate {
    key      = "tags"
    key_data = "db"
    type     = "equals"
    value = "rds"
  }
}

resource "opslevel_scorecard" "rds_scorecard" {
  name = "rds scorecard"
  description = "scorecard for apps using rds"
  owner_id = opslevel_team.platform_team.id
  filter_id = opslevel_filter.rds_filter.id
}

resource "opslevel_scorecard" "min_scorecard" {
  name = "minimal scorecard"
  owner_id = opslevel_team.platform_team.id
}
```

Step 2: terraform apply

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # opslevel_filter.rds_filter will be created
  + resource "opslevel_filter" "rds_filter" {
      + id           = (known after apply)
      + last_updated = (known after apply)
      + name         = "rds apps"

      + predicate {
          + key      = "tags"
          + key_data = "db"
          + type     = "equals"
          + value    = "rds"
        }
    }

  # opslevel_scorecard.min_scorecard will be created
  + resource "opslevel_scorecard" "min_scorecard" {
      + aliases        = (known after apply)
      + id             = (known after apply)
      + name           = "minimal scorecard"
      + owner_id       = (known after apply)
      + passing_checks = (known after apply)
      + service_count  = (known after apply)
      + total_checks   = (known after apply)
    }

  # opslevel_scorecard.rds_scorecard will be created
  + resource "opslevel_scorecard" "rds_scorecard" {
      + aliases        = (known after apply)
      + description    = "scorecard for apps using rds"
      + filter_id      = (known after apply)
      + id             = (known after apply)
      + name           = "rds scorecard"
      + owner_id       = (known after apply)
      + passing_checks = (known after apply)
      + service_count  = (known after apply)
      + total_checks   = (known after apply)
    }

  # opslevel_team.platform_team will be created
  + resource "opslevel_team" "platform_team" {
      + alias        = (known after apply)
      + id           = (known after apply)
      + last_updated = (known after apply)
      + name         = "platform"
    }

Plan: 4 to add, 0 to change, 0 to destroy.
```

Step 3: make modification to terraform file. `s/scorecard for apps using rds/this was changed in tf` and `s/minimal scorecard/min sc`

Step 4: apply again

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_scorecard.min_scorecard will be updated in-place
  ~ resource "opslevel_scorecard" "min_scorecard" {
        id             = "Z2lkOi8vb3BzbGV2ZWwvUnVicmljLzY2NQ"
      ~ name           = "minimal scorecard" -> "min sc"
        # (5 unchanged attributes hidden)
    }

  # opslevel_scorecard.rds_scorecard will be updated in-place
  ~ resource "opslevel_scorecard" "rds_scorecard" {
      ~ description    = "scorecard for apps using rds" -> "this was changed in rds"
        id             = "Z2lkOi8vb3BzbGV2ZWwvUnVicmljLzY2Ng"
        name           = "rds scorecard"
        # (6 unchanged attributes hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy.
```

Step 5: destroy

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # opslevel_filter.rds_filter will be destroyed
  - resource "opslevel_filter" "rds_filter" {
      - id   = "Z2lkOi8vb3BzbGV2ZWwvRmlsdGVyLzI1MzM" -> null
      - name = "rds apps" -> null

      - predicate {
          - key      = "tags" -> null
          - key_data = "db" -> null
          - type     = "equals" -> null
          - value    = "rds" -> null
        }
    }

  # opslevel_scorecard.min_scorecard will be destroyed
  - resource "opslevel_scorecard" "min_scorecard" {
      - aliases        = [
          - "min_sc",
        ] -> null
      - id             = "Z2lkOi8vb3BzbGV2ZWwvUnVicmljLzY2NQ" -> null
      - name           = "min sc" -> null
      - owner_id       = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS85MDA1" -> null
      - passing_checks = 0 -> null
      - service_count  = 283 -> null
      - total_checks   = 0 -> null
    }

  # opslevel_scorecard.rds_scorecard will be destroyed
  - resource "opslevel_scorecard" "rds_scorecard" {
      - aliases        = [
          - "rds_scorecard",
        ] -> null
      - description    = "this was changed in rds" -> null
      - filter_id      = "Z2lkOi8vb3BzbGV2ZWwvRmlsdGVyLzI1MzM" -> null
      - id             = "Z2lkOi8vb3BzbGV2ZWwvUnVicmljLzY2Ng" -> null
      - name           = "rds scorecard" -> null
      - owner_id       = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS85MDA1" -> null
      - passing_checks = 0 -> null
      - service_count  = 0 -> null
      - total_checks   = 0 -> null
    }

  # opslevel_team.platform_team will be destroyed
  - resource "opslevel_team" "platform_team" {
      - alias = "platform" -> null
      - id    = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS85MDA1" -> null
      - name  = "platform" -> null
    }

Plan: 0 to add, 0 to change, 4 to destroy.
```

**All CRUD operations succeeded!**

### CODE REVIEW

File: opslevel/resource_opslevel_scorecard.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+				Optional:    true,
+			},
+
+			// computed fields
+			"aliases": {
+				Type:        schema.TypeString,
+				Description: "The scorecard's aliases.",
+				Computed:    true,
+			},
+			"filter": {
```
Comment (taimoor-at-opslevel) [inline opslevel/resource_opslevel_scorecard.go:50]: @rocktavious `filter` and `owner` are full objects, so is `TypeString` appropriate here? Or do we need some kind of nested schema type?

Also, they are the same as `filterId`, `ownerId` except the full object. Should we still include them?

💭

Comment (rocktavious) [reply opslevel/resource_opslevel_scorecard.go:50]: We don't need them we only need to read the filter and owner id into the `filterId` and `ownerId` fields in the read function.

File: opslevel/resource_opslevel_scorecard.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+				Description: "The scorecard's filter.",
+				Computed:    true,
+			},
+			"owner": {
+				Type:        schema.TypeString,
+				Description: "The scorecard's owner.",
+				Computed:    true,
+			},
+			"passingChecks": {
+				Type:        schema.TypeString,
```
Comment (rocktavious) [inline opslevel/resource_opslevel_scorecard.go:61]: these should be int?

File: opslevel/resource_opslevel_scorecard.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func resourceScorecardUpdate(d *schema.ResourceData, client *opslevel.Client) error {
+	input := opslevel.ScorecardInput{}
+
+	if d.HasChange("name") {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_scorecard.go:136]: make sure to collapse this down into just building the input structure like the Infrastructure resource update function

File: Taskfile.yml [hunk 4]
```diff
@@ -104,6 +104,8 @@ tasks:
       FILES_TO_DELETE:
         sh: ls -1 "{{.WORKSPACE_DIR}}" | grep -v -e 'main.tf' -e 'backend.tf'
     cmds:
+      # TODO: this is broken because of line spaces! (the \n character)
```
Comment (taimoor-at-opslevel) [inline Taskfile.yml:107]: @davidbloss noticed this bug, could it be a difference in the shells we are using? I'm using zsh/oh-my-zsh and the macOS version of `ls`.

Comment (taimoor-at-opslevel) [reply Taskfile.yml:107]: Can we change the command such that it just deletes the `tfstate` file and backup file? I think it's better if the file doesn't touch any `.tf` files at all, the user can clean those manually if they want to.

**Review by rocktavious — APPROVED** [review]
Lovely!

**Pull request merged**